imap/imap.go

```diff
@@ -0,0 +1,630 @@
+				return text, nil
+			}
+			return text, err
+		case html.TextToken:
+			if !skip {
+				tmp := bytes.TrimSpace(z.Text())
+				if len(tmp) == 0 {
+					continue
+				}
+				tagText := make([]byte, len(tmp))
+				copy(tagText, tmp)
+				text = append(text, tagText)
+			}
+		case html.StartTagToken, html.EndTagToken:
+			tn, _ := z.TagName()
+			for _, nvTag := range nonVisibleTags {
+				if bytes.Equal(tn, nvTag) {
+					skip = (tt == html.StartTagToken)
+					break
+				}
+			}
+		}
+	}
+	return text, nil
+}
+
+// VisibleText will return any visible text from an HTML
+// email body.
+func (e *Email) VisibleText() ([][]byte, error) {
+	// if theres no HTML, just return text
+	if len(e.HTML) == 0 {
+		return [][]byte{e.Text}, nil
+	}
+	return VisibleText(bytes.NewReader(e.HTML))
+}
+
+// String is to spit out a somewhat pretty version of the email.
+func (e *Email) String() string {
+	return fmt.Sprintf(`
+----------------------------
+From:           %s
+To:             %s
+Internal Date:  %s
+Precedence:     %s
+Subject:        %s
+HTML:           %s
+
+Text:           %s
+----------------------------
+
+`,
+		e.From,
+		e.To,
+		e.InternalDate,
+		e.Precedence,
+		e.Subject,
+		string(e.HTML),
+		string(e.Text),
+	)
+}
+
+// Response is a helper struct to wrap the email responses and possible errors.
+type Response struct {
+	Email Email
+	Err   error
+}
+
+// newIMAPClient will initiate a new IMAP connection with the given creds.
+func newIMAPClient(info MailboxInfo) (*imap.Client, error) {
+	var client *imap.Client
+	var err error
+	if info.TLS {
+		client, err = imap.DialTLS(info.Host, new(tls.Config))
+		if err != nil {
+			return client, err
+		}
+	} else {
+		client, err = imap.Dial(info.Host)
+		if err != nil {
+			return client, err
+		}
+	}
+
+	_, err = client.Login(info.User, info.Pwd)
+	if err != nil {
+		return client, err
+	}
+
+	_, err = imap.Wait(client.Select(info.Folder, info.ReadOnly))
+	if err != nil {
+		return client, err
+	}
+
+	return client, nil
+}
+
+const dateFormat = "02-Jan-2006"
+
+// findEmails will run a find the UIDs of any emails that match the search.:
+func findEmails(client *imap.Client, search string, since *time.Time) (*imap.Command, error) {
+	var specs []imap.Field
+	if len(search) > 0 {
+		specs = append(specs, search)
+	}
+
+	if since != nil {
+		sinceStr := since.Format(dateFormat)
+		specs = append(specs, "SINCE", sinceStr)
+	}
+
+	// get headers and UID for UnSeen message in src inbox...
+	cmd, err := imap.Wait(client.UIDSearch(specs...))
+	if err != nil {
+		return &imap.Command{}, fmt.Errorf("uid search failed: %s", err)
+	}
+	return cmd, nil
+}
+
+var GenerateBufferSize = 100
+
+func generateMail(info MailboxInfo, search string, since *time.Time, markAsRead, delete bool) (chan Response, error) {
+	responses := make(chan Response, GenerateBufferSize)
+	client, err := newIMAPClient(info)
+	if err != nil {
+		close(responses)
+		return responses, fmt.Errorf("uid search failed: %s", err)
```

Comment:
Minor nit - Is this the right error here? The "uid search failed" threw me off, since this should just be connecting to the IMAP server.